### DISCUSSION TITLE
Source All Needs Assessment Summaries, Broken Down by Quarter & Region

### DISCUSSION THREAD

**Discussion by jtfairbank** [distributeaid/needs-assessment-storage#90] [Q&A] [answered]
Hey @coderbyheart, given the way Gatsby works and what we need for the landing site, it'd be easiest if we could source all needs assessment summaries broken down by quarter & region.  That'd limit the amount of data I'd have to hard-code to make the requests, and help future-proof the Gatsby import code if we add new regions, etc.

I'm imagining something like the overall summary query from #62, but the json results would be broken down in a way similar to this:

```
{
  '2022-q3': {
    'lesvos': {
      // same results as searching for quarter & region
      // https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?basicInfo.region=lesvos&timeOfYear.quarter=q3
      ...
    },
    ...
  },
  ...
}
```

Similarly if you request just 1 quarter it'd break the data down by region, or if you request just 1 region it'd bread the data down by quarter.

I'd recommend extending the existing API by adding a "details" parameter to access these breakdowns.

**Comment by coderbyheart** [comment]
Sourcing "all" needs assessment will be a problem, because those contain also responses for testing / preview forms.

How about the API can summarize multiple forms, however a list of **known good** IDs have to be provided, so the only thing that changes when a new needs assessment form is added is to add it's id to the query?

For example, instead of a single form summary using `https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary` the new endpoint `https://storage.needs-assessment.distributeaid.dev/forms/summary?ids=01G4X7XTMCD6MFHDT5KS7Z27GF,01GCW8TQFCE92TR1NXH8NEET7F` can be used to get a combined summary.

The endpoint would combine the data from these two forms, and add up the values for answers with the same ID. For example:

From [`01G4X7XTMCD6MFHDT5KS7Z27GF`](https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary): 

```json
"foodItems":{"rice":{"kg":248830}
```

From [01GCW8TQFCE92TR1NXH8NEET7F](https://storage.needs-assessment.distributeaid.dev/form/01GCW8TQFCE92TR1NXH8NEET7F/summary):

```json
"foodItems":{"rice":{"kg":3950},
```
Into:

```json
"foodItems":{"rice":{"kg":252780},
```

**Comment by jtfairbank** [comment, marked as answer]
Hey @coderbyheart!  I think we're talking about two different things here, let me clarify:

I want to be able to query the data for a single survey (e.g. `https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary`) but NOT have the data summed overall.  Instead, it'd be helpful to get each region's summary listed.  If there's only 1 survey per quarter then I don't need things broken down by quarter as well, though it'd be helpful to have the year / quarter listed in the `stats` attribute.  But if we reuse a survey for multiple quarters then I will need the data broken down by quarter as well.

Does that make sense?

It's less of an issue for me to request multiple surveys.  I can implement a whitelist of good survey ids on the gatsby side and loop through that, making 1 request per survey.

**Reply by coderbyheart** [reply]
Ah, so this form might have multiple quarters, but instead fetching these individually:

- https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?timeOfYear.quarter=q1
- https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?timeOfYear.quarter=q2
- https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?timeOfYear.quarter=q3

You would like to have the data in one response?

```
{
"q1": <Result of https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?timeOfYear.quarter=q1>,
"q2": <Result of https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?timeOfYear.quarter=q2>,
"q3": <Result of https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?timeOfYear.quarter=q3>
}
```

That can be solved by implementing a "group by" query: https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?groupBy=timeOfYear.quarter

Note that the form has no year associated to quarters. Quarters are `q1`, `q2`, `q3`, `q4`. So we don't know which year that refers to.

**Reply by jtfairbank** [reply]
Yes!  Very close.  So same thing with the regions as well.  Something like:

```
{
  "q1": {
    "southernGreece": <Result of https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?basicInfo.region=southernGreece&timeOfYear.quarter=q1>,
    "northernGreece": <Result of https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?basicInfo.region=northernGreece&timeOfYear.quarter=q1>,
    ...
  },
  "q2:" {
    "southernGreece": <Result of https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?basicInfo.region=southernGreece&timeOfYear.quarter=q2>,
    "northernGreece": <Result of https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?basicInfo.region=northernGreece&timeOfYear.quarter=q2>,
    ...
  },
  "q2:" {
    "southernGreece": <Result of https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?basicInfo.region=southernGreece&timeOfYear.quarter=q3>,
    "northernGreece": <Result of https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?basicInfo.region=northernGreece&timeOfYear.quarter=q3>,
    ...
  }
}
```

This will allow us to make 1 request (per survey) which only requires the survey code, and let's us present per-quarter / per-region data.  We can also present summary data by using GraphQL's aggregate functions on the gatsby side.  This means we can show current results per region, trends over time per quarter, summaries across surveys for the year, etc.

**Reply by coderbyheart** [reply]
Ok, I see, this is a second groupBy and can be done, too.

**Reply by jtfairbank** [reply]
Awesome!  Let me know if you need anything else from me on the needs assessment side.  I'll progress w/ the landing-site side of the implementation using a single survey for now.  :smile:

**Reply by coderbyheart** [reply]
It would be great to have the exact example you've built by hand just for comparison reasons.

**Reply by jtfairbank** [reply]
Like, a sample `json` response for a survey broken down the way I want it?  I can get that to you, probably tomorrow!

**Reply by coderbyheart** [reply]
Yes, that would be very helpful to ensure that we are on the same page.

**Reply by jtfairbank** [reply]
Hey @coderbyheart I did my best using form `01G4X7XTMCD6MFHDT5KS7Z27GF`.  It looks like it was only used once for Q3, but I included the group-by-quarter layer in case we use a form for multiple quarters, as well as a group-by-region layer. Note that regions that receive no responses are excluded ([I don't need empty responses like this](https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?basicInfo.region=paris&timeOfYear.quarter=q3)).

Github doesn't support json uploads so I added a `.txt` at the end to include the file directly in this discussion.

[sample-groupby-export-json-01G4X7XTMCD6MFHDT5KS7Z27GF.json.txt](https://github.com/distributeaid/needs-assessment-storage/files/9620964/sample-groupby-export-json-01G4X7XTMCD6MFHDT5KS7Z27GF.json.txt)

**Reply by coderbyheart** [reply]
Gist works good for things like this.
https://gist.github.com/coderbyheart/2bcc1ce13c03ee51529797e1750db6a6

**Reply by coderbyheart** [reply]
Thanks for the file!

**Reply by coderbyheart** [reply]
Fixed in 9c7dd82627c960d58be20d4e2bd3b36f6be4f6db

**Reply by coderbyheart** [reply]
Sample query: https://storage.needs-assessment.distributeaid.dev/form/01G4X7XTMCD6MFHDT5KS7Z27GF/summary?groupBy=timeOfYear.quarter,basicInfo.region

**Reply by jtfairbank** [reply]
Amazing, thank you so much @coderbyheart!
